### PR TITLE
Auto update of multi plotting

### LINKED ISSUE #25614: update multiplot
the multiplot GUI should update if the data is changed or deleted 

https://github.com/mantidproject/mantid/pull/24414/files#diff-59a3f22cce9ee9b0ef09d3d8db9c073dR24

### PR DESCRIPTION
**Pull request by AnthonyLim23**
Open elemental analysis
Make sure the search user directories includes the location of the data (I will email it)
Type 2695 into the box above load
Press load
If you tick all of the detectors a 2x2 plot will appear. 

If you run the below script a spike will appear in one of the plots

```
from mantid.simpleapi import *

name = "1_Total_2695"

ws0 = mtd[name]
ws = ws0
y= ws.dataY(0)
y[len(ws.dataX(0))-1] = 2000

AnalysisDataService.addOrReplace(name, ws)
```

If you delete one of the ws then the corresponding line will disappear from the plot
If you remove all the lines the subplot will close.

Fixes #25614. 

@NickDraper which section of the release notes do you think this is best suited?
---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/MultiPlotting/multi_plotting_context.py [hunk 1]
```diff
@@ -49,6 +50,12 @@ def add_annotate(self, subplotName, label):
     def add_vline(self, subplotName, xvalue, name):
         self.subplots[subplotName].add_vline(xvalue, name)
 
+    def removePlotLine(self, subplotName, name):
+        try:
+            self.subplots[subplotName].removePlotLine(name)
+        except:
```
Comment (Mantid-Matthew) [inline scripts/MultiPlotting/multi_plotting_context.py]: Would it be better to use a specific exception here, I think KeyError might do it?

File: scripts/MultiPlotting/subplot/subplot_ADS_observer.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from mantid.api import AnalysisDataServiceObserver
+
+
+class SubplotADSObserver(AnalysisDataServiceObserver):
+
+    def __init__(self, subplot):
+        super(SubplotADSObserver, self).__init__()
+        self._subplot = subplot
+
+        # self.observeClear(True)
```
Comment (Mantid-Matthew) [inline scripts/MultiPlotting/subplot/subplot_ADS_observer.py]: Does this want to un-commented or removed?

File: scripts/test/MultiPlotting/MultiPlottingContext_test.py [hunk 3]
```diff
@@ -6,10 +6,11 @@
 # SPDX - License - Identifier: GPL - 3.0 +
 import unittest
 
+from mantid.simpleapi import *
```
Comment (Mantid-Matthew) [inline scripts/test/MultiPlotting/MultiPlottingContext_test.py]: I've never quite been certain on this, is it our standard to import everything from the simpleapi as standard? I see this done very commonly but have always personally preferred importing just the required algorithms so it is easier to track down where they have come from in the code.

File: scripts/test/MultiPlotting/MultiPlottingContext_test.py [hunk 4]
```diff
@@ -70,5 +71,31 @@ def test_updateLayout(self):
             patch.assert_called_with(gridspec,figure,2)
 
 
+    def test_subplotEmptyTrue(self):
+        names = ["one","two","three"]
+        for name in names:
+            self.context.addSubplot(name, mock.Mock())
+
+        for name in names:
+             self.assertEquals(self.context.is_subplot_empty(name),True)
```
Comment (Mantid-Matthew) [inline scripts/test/MultiPlotting/MultiPlottingContext_test.py]: As a very minor point as of python 3.3 assertEquals is deprecated in favor of assertEqual for now both work equally well but might be worth using the singular for new code so as to me more forward compatible.

File: scripts/test/MultiPlotting/Subplot_test.py [hunk 5]
```diff
@@ -158,5 +160,50 @@ def test_addSubplot(self):
          self.subplot._context.update_gridspec.assert_called_with(4)
          self.assertEquals(self.subplot._update.call_count,1)
 
+    #def test_rm_ws_from_plots(self):
```
Comment (Mantid-Matthew) [inline scripts/test/MultiPlotting/Subplot_test.py]: stray comment

**Review by Matthew-Andrew — COMMENTED** [review]
Code looks good, a few minor comments below.

I will test the functionality once I am back a RAL.

**Review by Matthew-Andrew — COMMENTED** [review]
The script worked for me adding the line to the plot. Removing individual lines also removed them from the plot. I ran into a few related issues which are possibly beyond the scope of this PR however. 

  * If once removing all the lines from a plot you try and re-plot it you unsurprisingly get an error about being unable to find the relevant workspace in the ADS and the new plot is created blank. I don't think this is really an issue however as this is how I would expect it to behave in this case.
  * The second is that of after deleting some work-spaces you click load  the deleted work spaces are not reloaded in. I think it would be nice if clicking load reloaded all the workspaces you had deleted?

File: scripts/test/MultiPlotting/MultiPlottingContext_test.py (hunk 3 above)
Comment (AnthonyLim23) [reply scripts/test/MultiPlotting/MultiPlottingContext_test.py]: not needed, so deleted it

**Comment by AnthonyLim23** [comment]
I will have fixes pushed later today. Just want to test it first

**Review by Matthew-Andrew — APPROVED** [review]
LGTM :+1:

**Pull request merged**